mne/connectivity/cfc.py

```diff
@@ -0,0 +1,413 @@
+        this_env = amp_envelope_fa[trial]
+        amplitude_bin_means[trial] = [this_env[digitized[trial] == i].mean()
+                                      for i in range(1, len(phase_bins))]
+        if np.isnan(np.sum(amplitude_bin_means[trial])):
+            raise ValueError('Encountered nan when calculating mean'
+                             ' amplitude for bins.')
+
+        # Calculate normalized mean amplitude.
+        normalized_amplitudes[trial] = (amplitude_bin_means[trial] /
+                                        np.sum(amplitude_bin_means[trial]))
+        if np.round(normalized_amplitudes[trial].sum()) != 1:
+            raise AssertionError('Normalized amplitudes are incorrect')
+
+    return normalized_amplitudes, phase_bins
+
+
+@verbose
+def compute_phase_locked_stats(raw_spec_amplitudes, evoked_spec_amplitudes,
+                               freqs, trigger_inds, n_surrogates=10000,
+                               alpha=0.001, random_state=None, verbose=None):
+    """Compute a z-threshold and zscores for a phase peak-locked high amplitude.
+
+    Parameters
+    ----------
+    raw_spec_amplitudes : ndarray, shape (n_freqs, n_times)
+        Evoked spectral amplitude traces.
+    evoked_spec_amplitudes : ndarray, shape (n_freqs, n_times_epoch)
+        Normalized raw amplitude signals.
+    sfreq : float
+        Sampling frequency.
+    freqs : array, shape (n_freqs,)
+        Frequency points across range at which amplitudes are computed.
+    trigger_inds : array (trigger_inds,)
+        Indices of maxima in the phase signal.
+    n_surrogates : int
+        Number of surrogates. Default 10 ** 4.
+    alpha : float
+        Error rate allowed. Default 0.001.
+    random_state : None | int | instance of np.random.RandomState
+        np.random.RandomState to initialize the noise estimation.
+        Default value is None.
+    verbose : bool, str, int, or None
+        If not None, override default verbose level (see mne.verbose).
+
+
+    Returns
+    -------
+    spectral_amplitude_z : ndarray, shape (n_freqs, n_times)
+        Z-scored amplitude traces.
+    z_threshold : float
+        FDR corrected threshold for z-scored amplitude traces. If no
+        significant values are found, this will be np.nan.
+    """
+    from scipy.stats import norm
+
+    n_freqs = len(freqs)
+    n_times = raw_spec_amplitudes.shape[1]
+    epoch_length = evoked_spec_amplitudes.shape[1]
+    if n_freqs != raw_spec_amplitudes.shape[0]:
+        raise AssertionError('freqs and raw_spec_amplitudes differ' +
+                             ' in size on freq axis')
+
+    # permutation resampling for statistical significance:
+    # only need to compute mean and variance of surrogate distributions at one
+    # time point for all frequencies, since the surrogate trigger events could
+    # have occurred at any point
+    rng = check_random_state(random_state)
+    shifts = np.floor(rng.rand(2 * n_surrogates) * n_times)
```

Comment:
Hey Tom - thanks a lot for comments. I will address them if we think that
most of this code will stay the same. If we're going to significantly
change up this PR in order to use PACpy then I'll hold off.

On Wed, Oct 28, 2015 at 4:07 AM, Tom Dupré la Tour <notifications@github.com
> wrote:

> In mne/connectivity/cfc.py
> <https://github.com/mne-tools/mne-python/pull/2306#discussion_r43241810>:
>
> > +    """
> > +    from scipy.stats import norm
> > +
> > +    n_freqs = len(freqs)
> > +    n_times = raw_spec_amplitudes.shape[1]
> > +    epoch_length = evoked_spec_amplitudes.shape[1]
> > +    if n_freqs != raw_spec_amplitudes.shape[0]:
> > +        raise AssertionError('freqs and raw_spec_amplitudes differ' +
> > +                             ' in size on freq axis')
> > +
> > +    # permutation resampling for statistical significance:
> > +    # only need to compute mean and variance of surrogate distributions at one
> > +    # time point for all frequencies, since the surrogate trigger events could
> > +    # have occurred at any point
> > +    rng = check_random_state(random_state)
> > +    shifts = np.floor(rng.rand(2 * n_surrogates) * n_times)
>
> shifts = rng.randint(epoch_length + 1, n_times - epoch_length,
> n_surrogates)
> whould probably be faster and cleaner
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/mne-tools/mne-python/pull/2306/files#r43241810>.
>